### PR TITLE
[TASK] Refactor, testing sign-in/sign-up components. 

### PR DESCRIPTION
**Pull request by JimHolman1199**
**Refactoring sign-in and sign-up components.**
[1604](https://github.com/ita-social-projects/GreenCity/issues/1604);
[1605](https://github.com/ita-social-projects/GreenCity/issues/1605);

- Added reactive forms;

- Created new components and methods;

- Added unit testing;

- BugFix:

1.  'Sign up' button is enabled when password is invalid  [1540](https://github.com/ita-social-projects/GreenCity/issues/1540);

2.  'Successful Registration' message has a transparent background [1542](https://github.com/ita-social-projects/GreenCity/issues/1542);
3.  User can register with special characters and Cyrillic letters in 'User name' field [1512](https://github.com/ita-social-projects/GreenCity/issues/1512);
4. User cannot register when User name is 21-30 characters length [1483](https://github.com/ita-social-projects/GreenCity/issues/1483);
5. Wrong alignment of error message for 'User name' field [1484](https://github.com/ita-social-projects/GreenCity/issues/1484);
6. Tab is disabled for 'Sign in' link [1538](https://github.com/ita-social-projects/GreenCity/issues/1538);
7. Colors of 'forgot password' link on Sing in page does not match the design [1817](https://github.com/ita-social-projects/GreenCity/issues/817);
8. Bottom border of the 'Password' field is not displayed correctly [1194](https://github.com/ita-social-projects/GreenCity/issues/1194);
9. 'Sign in with Google' text on button has incorrect alignment [1555](https://github.com/ita-social-projects/GreenCity/issues/1555);
10. 'Picture' and 'Source' fields; 'Title' and 'Please select on or more tags' have got same locatet. [1200](https://github.com/ita-social-projects/GreenCity/issues/1200)
11. Fixed scroll on auth-modal component [1536](https://github.com/ita-social-projects/GreenCity/issues/1536);

### CODE REVIEW

**Review by romanovskyroma — APPROVED** [review]

**Review by vitaliipolitylo — APPROVED** [review]

**Review by skrebetsnataliia — APPROVED** [review]

**Review by luchward — APPROVED** [review]

File: src/app/component/auth/components/error/error.component.ts [hunk 1]
```diff
@@ -0,0 +1,15 @@
...
+
+@Component({
+  selector: 'app-error',
+  templateUrl: './error.component.html'
+})
+export class ErrorComponent implements OnInit {
+  @Input() errorMessage;
+  @Input() formElement;
+  ngOnInit():void {
+
```
Comment (mr-pavlenko) [inline src/app/component/auth/components/error/error.component.ts:12]: space

File: src/app/component/auth/components/sign-in/sign-in.component.ts [hunk 2]
```diff
@@ -152,5 +165,30 @@ export class SignInComponent implements OnInit, OnDestroy {
...
+  /**
+ * Returns the string according to the error that needs to be translated later.
+ *
+ * @param {obj} x FormControl on which the error occurred.
+ * @param {string} n Tag to know what element.
+ * @return {string} x Which needs to be translated.
+ */
+  public getErrorMessage({errors}: FormControl, tag: string){
+    // 
+    if(tag === 'email'){
```
Comment (mr-pavlenko) [inline src/app/component/auth/components/sign-in/sign-in.component.ts]: string move to constant

File: src/app/component/auth/components/google-btn/google-btn.component.scss [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+
+.google-text-sign-in {
+  flex: 1 1 auto;
+  align-self: center;
+  font-family: Roboto, sans-serif;
+  font-style: normal;
+  font-weight: normal;
+  font-size: 14px;
+  line-height: 16px;
+  letter-spacing: 0.21875px;
```
Comment (mr-pavlenko) [inline src/app/component/auth/components/google-btn/google-btn.component.scss:30]: Why this value so strange?

File: src/app/component/auth/components/sign-in/sign-in.component.scss [hunk 4]
```diff
@@ -186,6 +186,7 @@ $inputHeight : 48px;
       margin-bottom: 24px;
 
       .forgot-password {
+        color: #056B33;
```
Comment (mr-pavlenko) [inline src/app/component/auth/components/sign-in/sign-in.component.scss]: use variable

File: src/app/component/auth/components/sign-up/sign-up.component.ts [hunk 5]
```diff
@@ -12,52 +12,93 @@ import { UserOwnSignUp } from '@global-models/user-own-sign-up';
...
+              @Inject(SIGN_IN_TOKEN) private signInToken: ComponentType<any>) { }
 
   ngOnInit() {
+    this.InitFormReactive();
+    this.getFormFields();
     this.userOwnSignUp = new UserOwnSignUp();
     this.setNullAllMessage();
   }
 
+  public InitFormReactive() : void {
```
Comment (mr-pavlenko) [inline src/app/component/auth/components/sign-up/sign-up.component.ts]: naming 
MB OnFormInit ? InitReactiveForm ?

File: src/app/component/auth/components/sign-up/sign-up.component.ts [hunk 6]
```diff
@@ -176,4 +193,37 @@ export class SignUpComponent implements OnInit {
       panelClass: 'custom-dialog-container',
     });
   }
+  
+  public getErrorMessage({ errors }: FormControl, tag: string) {
```
Comment (mr-pavlenko) [inline src/app/component/auth/components/sign-up/sign-up.component.ts]: validators can be moved to a separate file.

Comment (mr-pavlenko) [reply src/app/component/auth/components/sign-up/sign-up.component.ts]: Or why It's public?

**Review by yulkats — APPROVED** [review]

File: src/app/component/auth/components/sign-up/sign-up.component.scss [hunk 7]
```diff
@@ -6,220 +6,186 @@ $title-font: 'Montserrat', sans-serif;
...
-
-  .main-image {
-    border-radius: 4px 0 0 4px;
-  }
 
-  .main-content-container {
-    margin: 0 55px;
-    width: 370px;
+    .under-error {
+      margin: 9px 0 4px;
```
Comment (mr-pavlenko) [inline src/app/component/auth/components/sign-up/sign-up.component.scss:42]: why 9 and 4 ?

File: src/app/component/auth/components/error/error.component.ts [hunk 8]
```diff
@@ -1,15 +1,37 @@
...
+          this.errorMessage = 'user.auth.sign-up.password-match';
+          break;
+        case 'minlength':
+          this.errorMessage = 'user.auth.sign-in.password-must-be-at-least-8-characters-long';
+          break;
+        case 'symbolInvalid':
+          this.errorMessage = this.controlName === 'password' 
+          ? 'user.auth.sign-up.password-symbols-error'
+          : 'user.auth.sign-up.user-name-size';
+          break;
```
Comment (mr-pavlenko) [inline src/app/component/auth/components/error/error.component.ts:34]: You can use strategy.

getErrorMsg = {
     required: function () {  this.errorMessage = 'user.auth.sign-in.field-is-required'; }
     email: function () {  this.errorMessage = 'user.auth.sign-in.this-is-not-email'; }
...
}

usage getErrorMsg['error'] ()

File: src/app/component/auth/components/error/error.component.spec.ts [hunk 9]
```diff
@@ -1,31 +1,51 @@
...
 import { async, ComponentFixture, TestBed } from '@angular/core/testing';
+import { FormControl, FormsModule, ReactiveFormsModule } from '@angular/forms';
+import { Component } from '@angular/core';
+import { ErrorComponent } from './error.component';
 
 
 describe('error component', () => {
-  let component: ErrorComponent;
-  let fixture: ComponentFixture<ErrorComponent>;
+  let testHostComponent: TestHostComponent;
```
Comment (mr-pavlenko) [inline src/app/component/auth/components/error/error.component.spec.ts]: Why did you use let?

File: src/app/component/auth/components/sign-up/sign-up.component.spec.ts [hunk 10]
```diff
@@ -34,22 +42,265 @@ describe('SignUpComponent', () => {
...
+    it('should call closeSignUpWindow ', () => {
+      // @ts-ignore
+      const spy = spyOn(component.matDialogRef, 'close').and.callThrough();
+      // @ts-ignore
+      component.closeSignUpWindow();
+      expect(spy).toHaveBeenCalled();
+    });
+
+    it('should call openSignUpPopup', () => {
+      // @ts-ignore
```
Comment (mr-pavlenko) [inline src/app/component/auth/components/sign-up/sign-up.component.spec.ts:93]: Why you need this ignore?

File: src/app/component/auth/components/google-btn/google-btn.component.scss (hunk 3 above)
Comment (JimHolman1199) [reply src/app/component/auth/components/google-btn/google-btn.component.scss:30]: Screenshot 2020-10-26 at 17 01 06
according to design

File: src/app/component/auth/components/sign-up/sign-up.component.spec.ts (hunk 10 above)
Comment (JimHolman1199) [reply src/app/component/auth/components/sign-up/sign-up.component.spec.ts:93]: > Why you need this ignore?

we use it to test private methods. are there other constructions for this?

File: src/app/component/auth/components/sign-up/sign-up.component.scss (hunk 7 above)
Comment (yulkats) [reply src/app/component/auth/components/sign-up/sign-up.component.scss:42]: according to design

File: src/app/component/auth/components/sign-up/sign-up.component.ts (hunk 6 above)
Comment (yulkats) [reply src/app/component/auth/components/sign-up/sign-up.component.ts]: fixed

File: src/app/component/auth/components/error/error.component.spec.ts (hunk 9 above)
Comment (JimHolman1199) [reply src/app/component/auth/components/error/error.component.spec.ts]: this is outdated code)

File: src/app/component/auth/components/error/error.component.ts (hunk 8 above)
Comment (JimHolman1199) [reply src/app/component/auth/components/error/error.component.ts]: fixed

File: src/app/component/auth/components/sign-up/sign-up.component.ts (hunk 5 above)
Comment (JimHolman1199) [reply src/app/component/auth/components/sign-up/sign-up.component.ts]: fixed

File: src/app/component/auth/components/sign-in/sign-in.component.scss (hunk 4 above)
Comment (JimHolman1199) [reply src/app/component/auth/components/sign-in/sign-in.component.scss]: fixed

File: src/app/component/auth/components/sign-in/sign-in.component.ts (hunk 2 above)
Comment (JimHolman1199) [reply src/app/component/auth/components/sign-in/sign-in.component.ts]: this is outdated code)

File: src/app/component/auth/components/error/error.component.ts (hunk 1 above)
Comment (JimHolman1199) [reply src/app/component/auth/components/error/error.component.ts:12]: fixed

**Pull request merged**